README.md
# tagmark (Python)

- [tagmark (Python)](#tagmark-python)
  - [Features And Demo](#features-and-demo)
  - [What is TagMark](#what-is-tagmark)
  - [Installation](#installation)
  - [Usage](#usage)
    - [sub command: convert](#sub-command-convert)
      - [Usage Examples](#usage-examples)
      - [Core Options Explantation and Design Details](#core-options-explantation-and-design-details)
    - [sub command: checktag](#sub-command-checktag)
      - [Usage Examples](#usage-examples-1)
    - [sub command: autotagdef](#sub-command-autotagdef)
      - [Usage Examples](#usage-examples-2)
    - [Condition File Details](#condition-file-details)
  - [Acknowledgments](#acknowledgments)
  - [Contributing](#contributing)
  - [TODO](#todo)

## Features And Demo

Here is a page as the demo of TagMark usage, which is all my bookmarks integrated in my blog:

<https://www.pwn.fan/tagmark/>

Fetures of the page:

- Substantial tag based bookmarks
  - 2500+ tagged bookmarks, which contain contents below:
    - cybersecurity
      - red team
      - blue team
      - etc
    - software development, devops, devsecops
    - blogs of a person, enterprise, team or organization which refer to the above topics
    - etc
  - 1700+ curated Github Repos
  - 950+ tags with detailed tag definitions
- Full featured tags
  - tag definitions (show/hide definition by left click on tags)
  - tag overview with counts
  - color difference depending on counts
- Simple but powerful header filter for each column
  - thick client: static, pure frontend and js based, so it's fast responding
  - simple and useful fitler grammer
  - quickly input tag name into filter by just a right click
  - press CTRL/CMD with left click in any filter input to call out multiple language document (English/Japanese/Chinses)
- Supporting for URL GET paramaters based filtering
  - static, pure frontend and js based
  - easy for sharing
- Columns related things
  - detailed Github repository information
  - suppressible columns

see [the gif demos on my blog](https://www.pwn.fan/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#what-is-tagmark).

I'm glad to share all my bookmarks with you. Welcome to help me enhance this /tagmark page in my blog by:

- [Contributing and Sharing Your Great Bookmark(s)](https://github.com/pwnfan/pwnfan.github.io/issues/1#issue-1729293100)
- [Reporting Improper Tag Definitions](https://github.com/pwnfan/pwnfan.github.io/issues/2#issue-1729295366)
- [Reporting Improper Tag For Bookmarks](https://github.com/pwnfan/pwnfan.github.io/issues/3#issue-1729507987)

## What is TagMark

TagMark is a tag based browser bookmark solution for developers.

I have written a blog post [TagMark: Maybe a Better Browser Bookmark Solution](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/) to introduce TagMark in details with a toc below:

- [What Is TagMark](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#what-is-tagmark)
- [What Is A "browser bookmark solution"](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#what-is-a-browser-bookmark-solution)
- [Why Do We Need Tags For Bookmarks](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#why-we-need-tags-for-bookmarks)
- [Why Not Diigo Bookmark Manager (Diigo My Library))](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#why-not-diigo-bookmark-manager-diigo-my-library)
- [TagMark's solution: How to build your own TagMark](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#tagmarks-solution-how-to-build-your-own-tagmark)
  - [Generating Your TagMark Data](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#generating-your-tagmark-data)
  - [Defining Your Tags With The Help Of ChatGPT](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#defining-your-tags-with-the-help-of-chatgpt)
  - [Periodically Upading Your TagMark Data and Tag Definitions](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#periodically-upading-your-tagmark-data-and-tag-definitions)
- [Advanced usages](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#advanced-usages)
- [FAQ](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#faq)

I recommend you read the blog first, to save time you can only read the [What is TagMark](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#what-is-tagmark) section and [TagMark's solution: How to build your own TagMark](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#tagmarks-solution-how-to-build-your-own-tagmark) section.

This repository `tagmark`(Python) is a part of the TagMark solution, and this doc is also only for the `tagmark`(Python) part.

## Installation

Requirements: Python>=3.11

**Step 1** install Python, we recommend installing in a virtual environment(pyenv/conda)

**Step 2** download the `tagmark-x.y.z-py3-none-any.whl` or `tagmark-x.y.z.tar.gz` package from the [Release Page](https://github.com/pwnfan/tagmark/releases)

**Step 3** install with pip:

```bash
pip install tagmark-x.y.z-py3-none-any.whl
```

or

```bash
pip install tagmark-x.y.z.tar.gz
```

**Step 4** Setup Github PAT (optional)

if you don't have any Github Repo Bookmarks, this step can be skipped.

1. [create a github personal access token(PAT)](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)
   `tagmark` requires PAT to access the Github API to get the repo info(stars, forks etc.) when a bookmark url is a Github repo url. The default settings to the PAT is recommended, which has no any priviledge for any action to any of your repos or settings.

2. create `.env` file containing the PAT. `.env` can be created in any directory you want:
    ```bash
    # create a file `.env` containing github aceess token
    echo "GITHUB_TOKEN=github_pat_XXX" > .env
    ```

**Step 5** run tagmark

command line options:

```bash
(testvenv) ➜ tagmark_cli --help
Usage: tagmark_cli [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  autotagdef  get tag definition automatically by ChatGPT
  checktag    check every tag has a definition
  convert     convert other bookmark formats into Tagmark format(json-lines)
```

## Usage

### sub command: convert

  </summary>

`convert` helps to convert other bookmark formats(diigo_chrome/tagmark_jsonlines) into Tagmark data format(json-lines).

```bash
(py311) ➜  tagmark_cli convert --help                                                            
Usage: tagmark_cli convert [OPTIONS]

  convert other bookmark formats into Tagmark format(json-lines)

Options:
  -i, --input-file-path FILE      input file path
  -f, --format [diigo_chrome|tagmark_jsonlines]
                                  format of the input file  [default:
                                  diigo_chrome]
  -o, --output-file-path FILE     output file path  [default:
                                  tagmark_ui_data.jsonl]
  -k, --keep_empty_keys BOOLEAN   whether keep keys with empty values
                                  [default: False]
  -c, --condition-json-path FILE  json file containing the condition for
                                  fitlering TagmarkItem  [default:
                                  /workspaces/tagmark-
                                  py/tagmark/condition_example.json]
  -b, --is-ban-condition BOOLEAN  If set to True, a TagmarkItem hits the
                                  `condition` will be banned, or it will be
                                  remained  [default: True]
  -t, --github_token TEXT         the GITHUB_TOKEN to access Github API,
                                  default will read from the .env file of the
                                  root dir of this project
  -u, --update-github-info-after-hours FLOAT
                                  update github info only when user specified
                                  number of hours has passed since the last
                                  update  [default: 23]
  -h, --help                      Show this message and exit.
```

Please refer to section [Core Options Explantation and Design Details](#core-options-explantation-and-design-details) for details of the options.

Please refer to [TagMark: Maybe a Better Browser Bookmark Solution > TagMark's solution: How to build your own TagMark > Generating Your TagMark Data](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#generating-your-tagmark-data) for usage scenarios of `convert` sub command.

#### Usage Examples

```bash
# convert from Diigo output(Chrome format) into TagMark json-lines format file tagmark_ui_data.jsonl
(py311) ➜  tagmark_cli convert -i data/16400249_chrome_2023_04_24_0a65e.html -f diigo_chrome

# actually not for conversion, but for refreshing Github repo information in the file tagmark_ui_data.jsonl
# and output into a new file new_tagmark_ui_data.jsonl, however if the one Github repo information has been 
# updated less than 12 hours, it will not be refreshed this time
(py311) ➜  tagmark_cli convert -i data/tagmark_ui_data.jsonl -f tagmark_jsonlines -o new_tagmark_ui_data.jsonl -u 12
```

#### Core Options Explantation and Design Details

- `-f`: the input file [default: diigo_chrome]
  - `diigo_chrome`: the Chrome format bookmarks exported from Diigo library, see [my blog](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#tagmarks-solution) for details
  - `tagmark_jsonlines`: indicate the input file is a tagmark format json-lines file, which is the output file of the `-o` option of this script. This may be confused, but it's not useless, the purpose of using a `-o` output file as the `-i` input file in a next run, is to refresh the Github info(Starts, Forks, Last Commit, etc) of the bookmarks who are Github repos.

- `-k`: whether keep keys with empty values [default: False]. This is to opitmize the output file size by removing the keys with empty values.

- `-c` and `-b`:

the whole workflow of tagmark:

```bash
                                            ┌──────────────┐
                                            │      -b      │
                                            │  This is a   │
                                            │ban-condition?│
                    ┌────────────┐          └───────┬──────┘
                    │TagmarkItem │                  │
                    │┌───────────┴┐                 ▼
┌───────────┐       └┤TagmarkItem │        ┌────────────────┐       ┌───────────┐
│    -i     │        │┌───────────┴┐──────>│Filter Condition│──────>│    -o     │
│Input File │───────>└┤TagmarkItem │       └────────────────┘       │Output File│
└───────────┘         │ ┌──────────┴─┐              ▲               └───────────┘
                      └─┤TagmarkItem │              │
                        │            │     ┌────────┴───────┐
                        └────────────┘     │       -c       │
                                           │ Condition File │
                                           └────────────────┘
```

`-c` specify a json file containing the condition for fitlering TagmarkItem, the default condition file is `tagmark/condition_example.json`, with content of:

```json
{
    "tags": ["Diigo"],
    "valid": true
}
```

What is a TagmarkItem? Taking a look at the output file format of `-o`, which is a json-lines format file, with one json data in one line. It is the json dump of TagmarkItem object, one line json data looks like:

```json
{
    "url": "https://github.com/jonschlinkert/remarkable",
    "id": 2,
    "valid": true,
    "title": "jonschlinkert/remarkable: Markdown parser, done right. Commonmark support, extensions, syntax plugins, high speed - all in one. Gulp and metalsmith plugins available. Used by Facebook, Docusaurus and many others! Use https://github.com/breakdance/breakdan",
    "tags": ["dev", "frontend", "javascript", "markdown"],
    "is_github_url": true,
    "github_repo_info": {
        "url": "https://github.com/jonschlinkert/remarkable",
        "owner": "jonschlinkert",
        "name": "remarkable",
        "description": "Markdown parser, done right. Commonmark support, extensions, syntax plugins, high speed - all in one. Gulp and metalsmith plugins available. Used by Facebook, Docusaurus and many others! Use https://github.com/breakdance/breakdance for HTML-to-markdown conversion. Use https://github.com/jonschlinkert/markdown-toc to generate a table of contents.",
        "time_created": "2014-09-01T17:57:42Z",
        "time_last_commit": "2023-03-30T05:55:40Z",
        "count_star": 5514,
        "count_fork": 396,
        "count_watcher": 5514,
        "topics": [
            "commonmark",
            "compile",
            "docusaurus",
            "gfm",
            "javascript",
            "jonschlinkert",
            "markdown",
            "markdown-it",
            "markdown-parser",
            "md",
            "node",
            "nodejs",
            "parse",
            "parser",
            "syntax-highlighting"
        ]
    },
    "time_added": "1682907038"
}
```

you can treat this json structure as the data structure of a TagmarkItem, `-c` condition file and `-b` specify a fitler telling tagmark if or not to output a TagmarkItem into the `-o` output file.

for example, if you _**do not**_ need any lines with tag `javascript` _**or**_ `css` to be output in the output file, you should specify your condition file by `-c my_condition.json` with the content below:

```json
{
    "tags": ["javascript", "css"],
}
```

and you need to specify the `-b True (default)` option, which means if a TagmarkItem meets the condition, it will be banned and will not be exported into to output file.

On the contrary, if you only need lines with tag `javascript` _**or**_ `css` to be output into the output file, you need to specify the `-b False` option, which means if a TagmarkItem meets the condition, it will be picked out(not banned) and put into the output file.

Check the [Condition File Details](#condition-file-details) section for more about condition file.

### sub command: checktag


`checktag` helps to verify if every tag in the output file generated by the `-o` parameter of the `convert` sub command has a definition in the tag definition file. This ensures the web UI [tagmark-ui](https://github.com/pwnfan/tagmark-ui) works better.

```bash
(py311) ➜  tagmark_cli checktag --help
Usage: tagmark_cli checktag [OPTIONS]

  check every tag has a definition

Options:
  -t, --tagmark-data-file-path FILE
                                  the tagmark data file path, which may be the
                                  output file generated by the `-o` parameter
                                  of the `convert` sub command
  -d, --tag-definition-file-path FILE
                                  tag definition file path
  -c, --condition-json-path FILE  json file containing the condition for
                                  fitlering TagmarkItem, here only the value
                                  of `tags` field in the file will be used,
                                  and this condition must be a ban condition
                                  [default: /Users/pwnfan/Desktop/Projects/tag
                                  mark/tagmark/condition_example.json]
  -a, --add-no-def-tag BOOLEAN    if set to `True`, a new tags definition file
                                  will be generated, which includes tags in
                                  the old tag definition file, and tags in the
                                  tagmark data file(specified by -t) without a
                                  tag definition. The latter tags are added as
                                  keys with a placeholder value specified by
                                  `-p / --no-def-tag-value-placeholder`
                                  parameter. If set to `False` and any tag
                                  without definition exists, an error will be
                                  raised  [default: True]
  -p, --no-def-tag-value-placeholder TEXT
                                  explained in the `-a` parameter  [default:
                                  !!!NO DEFINITION FOR THIS TAG, PLEASE ADD
                                  HERE!!!]
  --help                          Show this message and exit.
```

Please refer to

- [TagMark: Maybe a Better Browser Bookmark Solution > TagMark's solution: How to build your own TagMark > Defining Your Tags With The Help Of ChatGPT](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#defining-your-tags-with-the-help-of-chatgpt)
- [TagMark: Maybe a Better Browser Bookmark Solution > TagMark's solution: How to build your own TagMark > Periodically Upading Your TagMark Data and Tag Definitions](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#periodically-upading-your-tagmark-data-and-tag-definitions)

for usage scenarios of `checktag` sub command.
#### Usage Examples

```bash
# check every tag in `tagmark_ui_data.jsonl` has a definition in `tag_definitions.json`. If not so a new tag definition file whose name like `new-20230519183413-tag_definitions.json` will be generated, in this file tags already have definition will be added, as well as tags without definition, the latter tags will be added as keys with a placeholder value(default) "!!!NO DEFINITION FOR THIS TAG, PLEASE ADD HERE!!"
tagmark_cli checktag -t tagmark_ui_data.jsonl -d tag_definitions.json -c my-condition.json
```

```bash
# check every tag in `tagmark_ui_data.jsonl` has a definition in `tag_definitions.json`. If not error `tagmark.cli.TagDefinitionMissingError: tags without definition found` will be raise, with the details of the missing definition tags.
tagmark_cli checktag -t tagmark_ui_data.jsonl -d tag_definitions.json -c my-condition.json -a false
```

### sub command: autotagdef

`autotagdef` helps to get tag definition automatically from ChatGPT **based on the prompt given by users.**

```bash
(py311) ➜  tagmark git:(dev) ✗ tagmark_cli autotagdef --help
Usage: tagmark_cli autotagdef [OPTIONS]

  get tag definition automatically by ChatGPT

Options:
  -d, --tag-definition-file-path FILE
                                  tag definition file path
  -f, --gpt-prompt-ending-flag TEXT
                                  the ending flag of the prompt(question) sent
                                  to ChatGPT to get a answer about a tag
                                  definiton  [default: ?]
  -c, --gpt-config-file-path FILE
                                  the config file for invoking ChatGPT API, we
                                  sugguest setting `access_token` in the
                                  config file, see https://github.com/acheong0
                                  8/ChatGPT#--optional-configuration for
                                  details.
  -i, --gpt-conversation-id TEXT  the id of conversation in which to (continue
                                  to) interact with ChatGPT, if set to `None`
                                  a new conversation will be created. See http
                                  s://github.com/acheong08/ChatGPT/wiki/V1#ask
                                  for details.
  -t, --gpt-timeout INTEGER       the timeout that GPT answers one
                                  question(get one tag definition)  [default:
                                  60]
  -p, --no-def-tag-value-placeholder TEXT
                                  explained in the `-a` parameter  [default:
                                  !!!NO DEFINITION FOR THIS TAG, PLEASE ADD
                                  HERE!!!]
  --help                          Show this message and exit.
```

Please refer to

- [TagMark: Maybe a Better Browser Bookmark Solution > TagMark's solution: How to build your own TagMark > Defining Your Tags With The Help Of ChatGPT](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#defining-your-tags-with-the-help-of-chatgpt)
- [TagMark: Maybe a Better Browser Bookmark Solution > TagMark's solution: How to build your own TagMark > Periodically Upading Your TagMark Data and Tag Definitions](https://pwnfan.github.io/en/post/TagMark-Maybe-a-Better-Browser-Bookmark-Solution/#periodically-upading-your-tagmark-data-and-tag-definitions)

for usage scenarios of `autotagdef` sub command.

#### Usage Examples

```bash
poetry run tagmark_cli autotagdef -d tag_definitions.json -c gpt_config.json -i cca29e06-692e-4a52-84be-c47a7ca524cc
```


```bash
(py311) ➜  tagmark git:(dev) ✗ poetry run tagmark_cli autotagdef -d tag_definitions.json -c gpt_config.json -i cca29e06-692e-4a52-84be-c47a7ca524cc
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 471/471 [00:45<00:00, 10.41it/s]
{"filename": "cli.py", "func_name": "autotagdef", "level": "info", "lineno": 345, "logger": "tagmark.cli", "msg": "new tags definition file has been generated", "new_tags_definition_file": "PosixPath('/Users/pwnfan/Desktop/Projects/tagmark/new-20230522223828-tag_definitions.json')", "process": 82747, "process_name": "MainProcess", "timestamp": "2023-05-22T22:38:28.280672"}
{"auto_tag_make_stats": "AutoTagMakeStats(count_total_tags=471, count_auto_made_success=1, count_auto_made_fail=0, count_no_prompt=0, count_already_defined=470)", "filename": "cli.py", "func_name": "autotagdef", "level": "info", "lineno": 349, "logger": "tagmark.cli", "msg": "statistics", "process": 82747, "process_name": "MainProcess", "timestamp": "2023-05-22T22:38:28.334363"}
```

### Condition File Details

Note that not all keys in TagmarkItem are supported in condition filter files, here is a table for details:

| key              | value type    | supported in condition file | condition example                  | meaning                                                |
|------------------|---------------|:---------------------------:|------------------------------------|--------------------------------------------------------|
| url              | string        |             yes             | "url": ["github", "stackoverflow"] | url contains "github" _**or**_ "stackoverflow"         |
| id               | int           |             no              | -                                  | -                                                      |
| valid            | boolean       |             yes             | "valid": true                      | the url is valid(valid check haven't been implemented) |
| title            | string        |             yes             | (similar to `url`)                 | (similar to `url`)                                     |
| tags             | array         |             yes             | "tags": ["python", "javascript"]   | tags contains "python" _**or**_ "javascript"           |
| is_github_url    | boolean       |             yes             | (similar to `valid`)               | (similar to `valid`)                                   |
| github_repo_info | nested object |             no              | -                                  | -                                                      |
| time_added       | string        |             no              | -                                  | -                                                      |

All values in condition file is **case-sensitive**.

## Acknowledgments

- [tqdm](https://github.com/tqdm/tqdm)
- [revchatgpt](https://github.com/acheong08/ChatGPT)

CHANGELOG.md
# Changelog

Order: The higher up, the newer.

## v0.2.0

* e400bc8 infra.build>modify.new: bump requests from 2.28.2 to 2.31.0 to fix security vul reported by github dependabot
* 9d8c6d9 docs>new: add new content into README
* f780edd code.misc>modify.misc: fix word spelling of help msg in convert sub command
* cb7b733 code.tool>modify.new: add output file log info into `convert`subcommand
* cac1bda code.misc>new: add sub command "autotagdef"
* e4e06b0 code.misc>modify.misc: add comment to `checktag` sub command code
* 7fc0967 docs>new: add new content into README
* a72d352 code.misc>modify.refactor: change defualt value of "checktag -c",  and some help message modify
* 112eeda code.misc>new: add msg to show rate of process in `convert` command
* 5199240 code.misc>new: Tagmark.get_github_repo_infos add condition filter
* c07d897 code.core>modify.style: add support for utf-8 encoding in output jsonlines file
* 54a19bc code.misc>new: add sub command "checktag"
* 9ef2cd6 docs>modify.style: fix white-space in README
* e4e18eb docs>new: add new TODO into README
* f5c4bb6 docs>new: add new section into CONTRIBUTING
* 553f447 docs>new: add new TODO into README
* dca3576 code.tests>new: add test case for Tagmark.count_github_url

## v0.1.0

* finished basic features


tagmark/cli.py
import json
import os
import warnings
from datetime import datetime
from pathlib import Path

import click
from dotenv import find_dotenv, load_dotenv

from tagmark.core.convert import BaseConverter
from tagmark.core.log import LogHandler, LogLevel, get_level_logger
from tagmark.tools.autotagdef import AutoTagDefinitionMarker, AutoTagMakeStats
from tagmark.tools.checktag import TagsChecker
from tagmark.tools.convert import diigo
from tagmark.tools.convert import tagmark as tagmark_convert

load_dotenv(dotenv_path=find_dotenv(usecwd=True))


class TagDefinitionUnusedWarning(Warning):
    pass


class TagDefinitionMissingError(Exception):
    pass


__logger = get_level_logger(
    name="tagmark.cli",
    level=LogLevel.INFO,
    handlers=[
        LogHandler.CONSOLE,
    ],
)
__logger.bind(scope="cli")


@click.group(context_settings=dict(help_option_names=["-h", "--help"]))
def cli():
    pass


@cli.command(
    context_settings=dict(help_option_names=["-h", "--help"]),
    no_args_is_help=True,
    help="convert other bookmark formats into Tagmark format(json-lines)",
)
@click.option(
    "-i",
    "--input-file-path",
    type=click.Path(exists=True, file_okay=True, dir_okay=False),
    help="input file path",
)
@click.option(
    "-f",
    "--format",
    type=click.Choice(["diigo_chrome", "tagmark_jsonlines"]),
    default="diigo_chrome",
    show_default=True,
    show_choices=True,
    help="format of the input file",
)
@click.option(
    "-o",
    "--output-file-path",
    type=click.Path(exists=False, file_okay=True, dir_okay=False),
    default="tagmark_ui_data.jsonl",
    show_default=True,
    help="output file path",
)
@click.option(
    "-k",
    "--keep_empty_keys",
    type=click.BOOL,
    default=False,
    show_default=True,
    help="whether keep keys with empty values",
)
@click.option(
    "-c",
    "--condition-json-path",
    type=click.Path(exists=True, file_okay=True, dir_okay=False),
    default=Path(__file__).absolute().parent.joinpath("condition_example.json"),
    show_default=True,
    help="json file containing the condition for fitlering TagmarkItem",
)
@click.option(
    "-b",
    "--is-ban-condition",
    type=click.BOOL,
    default=True,
    show_default=True,
    help="If set to True, a TagmarkItem hits the `condition` will be banned, or it will be remained",
)
@click.option(
    "-t",
    "--github_token",
    type=str,
    default=os.environ.get("GITHUB_TOKEN"),
    show_default=False,
    help="the GITHUB_TOKEN to access Github API, default will read from the .env file of the root dir of this project",
)
@click.option(
    "-u",
    "--update-github-info-after-hours",
    type=float,
    default=23,
    show_default=True,
    help="update github info only when user specified number of hours has passed since the last update",
)
def convert(
    input_file_path: str,
    format: str,
    output_file_path: str,
    keep_empty_keys: bool,
    condition_json_path: str,
    is_ban_condition: bool,
    github_token: str,
    update_github_info_after_hours: float,
):
    converter: BaseConverter = None
    match format:
        case "diigo_chrome":
            converter = diigo.ChromeConverter()
        case "tagmark_jsonlines":
            converter = tagmark_convert.JsonLinesConverter()
        case _:
            raise ValueError(f"unsupported format: {format}")
    _items: list[dict] = converter.load_original_items(
        data_source=Path(input_file_path)
    )
    converter.convert_to_tagmark(items=_items)
    with open(Path(condition_json_path), "r") as _f:
        condition: dict = json.load(_f)
        converter.tagmark.get_github_repo_infos(
            access_token=github_token,
            condition=condition,
            after_hours=update_github_info_after_hours,
            is_ban_condition=is_ban_condition,
        )
        converter.tagmark.dump_to_json_lines(
            output_path=Path(output_file_path),
            keep_empty_keys=keep_empty_keys,
            condition=condition,
            is_ban_condition=is_ban_condition,
        )
        __logger.info(
            msg="new tagmark data file has been generated",
            tagmark_data=Path(output_file_path).absolute(),
        )


@cli.command(
    context_settings=dict(help_option_names=["-h", "--help"]),
    no_args_is_help=True,
    help="check every tag has a been defined",
)
@click.option(
    "-d",
    "--tagmark-data-json-path",
    type=click.Path(exists=True, file_okay=True, dir_okay=False),
    help="the tagmark data json file path, which may be the output file generated by the `-o` parameter of the `convert` sub command",
)
@click.option(
    "-t",
    "--tags-json-path",
    type=click.Path(exists=True, file_okay=True, dir_okay=False),
    help="tags.json file path",
)
@click.option(
    "-c",
    "--condition-json-path",
    type=click.Path(exists=True, file_okay=True, dir_okay=False),
    default=Path(__file__).absolute().parent.joinpath("condition_example.json"),
    show_default=True,
    help="json file containing the condition for filtering TagmarkItem, here only the value of `tags` field in the file will be used, and this condition must be a ban condition",
)
@click.option(
    "-a",
    "--add-new-tags",
    type=bool,
    default=True,
    show_default=True,
    help="if set to `True`, a new tags.json file will be generated, which includes old tags in tag.json file, and new tags in the tagmark data file(specified by -t).",
)
def checktag(
    tagmark_data_json_path: str,
    tags_json_path: str,
    condition_json_path: str,
    add_new_tags: bool,
):
    tags_checker: TagsChecker = TagsChecker(
        tagmark_data_json_path=tagmark_data_json_path,
        tags_json_path=tags_json_path,
        condition_json_path=condition_json_path,
    )
    tags_checker.check_tags()
    __logger.info(tags_count=tags_checker.check_result.count)

    if tags_checker.check_result.tags_only_in_tags_json:
        __logger.warn(
            msg="tags only in tags.json(unused tags) found",
            tags_unused=tags_checker.check_result.tags_only_in_tags_json,
        )
        warnings.warn(
            message=f"tags only in tags.json(unused tags) found: {sorted(tags_checker.check_result.tags_only_in_tags_json)}",
            category=TagDefinitionUnusedWarning,
        )

    if tags_checker.check_result.tags_only_in_data_source:
        __logger.warn(
            msg="tags only in data source(undefined tags) found",
            tags_only_in_data_source=sorted(
                tags_checker.check_result.tags_only_in_data_source
            ),
        )

        if not add_new_tags:
            raise TagDefinitionMissingError(
                f"tags only in data source(undefined tags) found: {sorted(tags_checker.check_result.tags_only_in_data_source)}"
            )
        else:
            tags_json_path: Path = Path(tags_json_path)
            new_tags_json_file: Path = tags_json_path.parent.joinpath(
                f"new-{datetime.now().strftime('%Y%m%d%H%M%S')}-{tags_json_path.name}"
            )
            tags_checker.generate_new_tag_infos()
            with open(new_tags_json_file, "w") as f:
                json.dump(
                    obj=tags_checker.new_tag_infos,
                    fp=f,
                    sort_keys=True,
                    indent=4,
                    ensure_ascii=False,
                )
                __logger.info(
                    msg="new tags definition file has been generated",
                    new_tags_definition_file=new_tags_json_file.absolute(),
                )


@cli.command(
    context_settings=dict(help_option_names=["-h", "--help"]),
    no_args_is_help=True,
    help="get tag definition automatically by ChatGPT",
)
@click.option(
    "-d",
    "--tags-info-json-path",
    type=click.Path(exists=True, file_okay=True, dir_okay=False),
    help="tags.json(tags information) file path",
)
@click.option(
    "-c",
    "--gpt-config-file-path",
    type=click.Path(exists=True, file_okay=True, dir_okay=False),
    default="~/.config/revChatGPT/config.json",
    help="the config file for invoking ChatGPT API, we sugguest setting `access_token` in the config file, see https://github.com/acheong08/ChatGPT#--optional-configuration for details.",
)
@click.option(
    "-i",
    "--gpt-conversation-id",
    type=str,
    default=None,
    show_default=True,
    help="the id of conversation in which to (continue to) interact with ChatGPT, if set to `None` a new conversation will be created. See https://github.com/acheong08/ChatGPT/wiki/V1#ask for details.",
)
@click.option(
    "-t",
    "--gpt-timeout",
    type=int,
    default=60,
    show_default=True,
    help="the timeout that GPT answers one question(get one tag definition)",
)
@click.option(
    "-l",
    "--little-info-tag-is-ok",
    type=click.BOOL,
    default=False,
    show_default=True,
    help="",
)
def autotagdef(
    tags_info_json_path,
    gpt_config_file_path,
    gpt_conversation_id,
    gpt_timeout,
    little_info_tag_is_ok,
):
    # load tag definitions from file
    _tags_info_json_path: Path = Path(tags_info_json_path)
    with open(Path(_tags_info_json_path)) as _f:
        _tags_info: dict[str:dict] = json.load(_f)

    # loag gpt config from file
    with open(Path(gpt_config_file_path)) as _f:
        _gpt_config = json.load(_f)

    # auto make tag definitions
    _auto_tag_def_maker: AutoTagDefinitionMarker = AutoTagDefinitionMarker(
        gpt_config=_gpt_config,
        conversation_id=gpt_conversation_id,
        timeout=gpt_timeout,
    )

    _new_tag_definitions: dict[str:str]
    _auto_tag_make_stats: AutoTagMakeStats
    (
        _new_tag_definitions,
        _auto_tag_make_stats,
    ) = _auto_tag_def_maker.auto_define_tags(
        tag_definitions=_tags_info,
        little_info_tag_is_ok=little_info_tag_is_ok,
    )

    new_tags_info: dict[str:dict] = _tags_info.copy()
    for _tag, _definition in _new_tag_definitions.items():
        new_tags_info[_tag]["definition"] = _definition

    # write new tag definitions into file
    new_tags_info_json_path: Path = _tags_info_json_path.parent.joinpath(
        f"new-{datetime.now().strftime('%Y%m%d%H%M%S')}-{_tags_info_json_path.name}"
    )
    with open(new_tags_info_json_path, "w") as f:
        json.dump(
            obj=new_tags_info,
            fp=f,
            sort_keys=True,
            indent=4,
            ensure_ascii=False,
        )
        __logger.info(
            msg="new tags.json has been generated",
            new_tags_definition_file=new_tags_info_json_path.absolute(),
        )
        __logger.info(
            msg="statistics",
            auto_tag_make_stats=_auto_tag_make_stats,
        )


if __name__ == "__main__":
    cli()
